src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs

```diff
@@ -0,0 +1,213 @@
+{
+    internal static class ActivityExtensions
+    {
+        internal static IEnumerable<OtlpTrace.ResourceSpans> ToOtlpResourceSpans(this IEnumerable<Activity> activityBatch)
+        {
+            var resourceToLibraryAndSpans = GroupByResourceAndLibrary(activityBatch);
+            var resourceSpansList = new List<OtlpTrace.ResourceSpans>(resourceToLibraryAndSpans.Count);
+
+            foreach (var resource in resourceToLibraryAndSpans)
+            {
+                var libraryList = new List<OtlpTrace.InstrumentationLibrarySpans>(resource.Value.Count);
+                foreach (var activitySourceEntry in resource.Value)
+                {
+                    var activitySource = activitySourceEntry.Key;
+                    var otlpLibrarySpans = new OtlpTrace.InstrumentationLibrarySpans
+                    {
+                        InstrumentationLibrary = new OtlpCommon.InstrumentationLibrary
+                        {
+                            Name = activitySource.Name, // Name is enforced to not be null, but it can be empty.
+                            Version = activitySource.Version ?? string.Empty, // NRE throw by proto
+                        },
+                    };
+
+                    otlpLibrarySpans.Spans.AddRange(activitySourceEntry.Value);
+                    libraryList.Add(otlpLibrarySpans);
+                }
+
+                var otlpResources = new OtlpResource.Resource();
+                otlpResources.Attributes.AddRange(
+                    resource.Key.Attributes.Select(ToOtlpAttribute));
+
+                var otlpResourceSpans = new OtlpTrace.ResourceSpans
+                {
+                    Resource = otlpResources,
+                };
+                otlpResourceSpans.InstrumentationLibrarySpans.AddRange(libraryList);
+
+                resourceSpansList.Add(otlpResourceSpans);
+            }
+
+            return resourceSpansList;
+        }
+
+        internal static OtlpTrace.Span ToOtlpSpan(this Activity activity)
+        {
+            // protobuf doesn't understand Span<T> yet: https://github.com/protocolbuffers/protobuf/issues/3431
+            Span<byte> traceIdBytes = stackalloc byte[16];
+            Span<byte> spanIdBytes = stackalloc byte[8];
+
+            activity.TraceId.CopyTo(traceIdBytes);
+            activity.SpanId.CopyTo(spanIdBytes);
+
+            var parentSpanIdString = ByteString.Empty;
+            if (activity.ParentSpanId != default)
+            {
+                Span<byte> parentSpanIdBytes = stackalloc byte[8];
+                activity.ParentSpanId.CopyTo(parentSpanIdBytes);
+                parentSpanIdString = ByteString.CopyFrom(parentSpanIdBytes.ToArray());
+            }
+
+            var startTimeUnixNano = activity.StartTimeUtc.ToUnixTimeNanoseconds();
+            var otlpSpan = new OtlpTrace.Span
+            {
+                Name = activity.OperationName, // TODO: map DisplayName to anything?
```

Comment:
In principle, I'm assuming that we are not going to use `DisplayName`.

Reply:
I think the goal is to use DisplayName everywhere. OperationName was part of Activity already and it was not "settable" after construction. The SpanName in OT needs to be settable afterwards, which is why Activity introduced DisplayName.

In short, always use DisplayName and forget about OperationName. (DisplayName gets the OperationName by default. DisplayName can be changed, but OperationName cannot be)